File: 0435_non_overlapping_intervals/Solution.swift
// LeetCode 0435 - Non-overlapping Intervals
// https://leetcode.com/problems/non-overlapping-intervals/

class Solution {
    func eraseOverlapIntervals(_ intervals: [[Int]]) -> Int {
        let sorted = intervals.sorted { $0[1] < $1[1] }
        var removed = 0
        var end = Int.min

        for interval in sorted {
            let start = interval[0]
            let finish = interval[1]
            if start < end {
                removed += 1
            } else {
                end = finish
            }
        }

        return removed
    }
}

File: 0435_non_overlapping_intervals/solution.rb
# LeetCode 0435 - Non-overlapping Intervals
# https://leetcode.com/problems/non-overlapping-intervals/

class Solution
  def erase_overlap_intervals(intervals)
    sorted = intervals.sort_by { |interval| interval[1] }
    removed = 0
    end_time = -Float::INFINITY

    sorted.each do |start, finish|
      if start < end_time
        removed += 1
      else
        end_time = finish
      end
    end

    removed
  end

  alias_method :eraseOverlapIntervals, :erase_overlap_intervals
end
